docs/src/explanations/naming.md

```diff
@@ -0,0 +1,214 @@
+---
+layout: docs
+title:  "Naming"
+section: "chisel3"
+---
+
+Historically, Chisel has had trouble reliably capturing the names of signals. The reasons for this are due to (1)
+primarily relying on reflection to find names, (2) using `@chiselName` macro which had unreliable behavior, and (3)
+when all else fails using the `.suggestName("foo")` API.
+
+With Chisel 3.4 and later, a custom Scala compiler plugin was introduced which enabled reliably and automatically
+capturing the names of signals, when they are declared. In addition, this release included prolific use of a new
+prefixing API which enabled more stable naming of signals programmatically generated from function calls.
+
+This document explains how naming now works in Chisel for signal and module names. For cookbook examples on how to fix
+systemic name-stability issues, please refer to the naming [cookbook](../cookbooks/naming.md).
+
+### Compiler Plugin
+
+```scala mdoc:invisible
+import chisel3.plugin._
+import chisel3._
+import chisel3.experimental.prefix
+import chisel3.experimental.noPrefix
+import chisel3.stage.ChiselStage
```

Comment:
I'm not sure this should be invisible, or at least, not all of it. We talk about using `prefix` and `noPrefix` below but then hide the imports to use them.